File: src/native/ping_pong_erl/ping_pong.erl
%% ping_pong.erl
%%
%% This module implements a distributed message-passing system using the Actor model in Erlang.
%% It simulates clients sending messages (pings) to the server, which processes messages
%% and responds with pongs. The main process ensures orderly termination.
%%
%% Parameters:
%% - NumClients: Number of client processes.
%% - NumMessages: Number of messages each client sends.
%% - verbose: Verbose mode for debugging.
%%
%% The system terminates when all clients and the server finish processing.

-module(ping_pong).
-export([main/0, main/1]).

% Entry point of the program
main() ->
    start(4, 10, false).

main(Args) ->
    case parse_args(Args) of
        {ok, NumClients, NumMessages, Verbose} ->
            start(NumClients, NumMessages, Verbose); % Start the system with parsed parameters
        {error, Msg} ->
            io:format("Error: ~s~nUsage: main <num_clients> <num_messages> [verbose]~n", [Msg])
    end.

% Initializes the system: spawns server and clients
start(NumClients, NumMessages, Verbose) ->
    Main = self(), % Store reference to the main process

    % Spawn server processes
    io:format("Starting server...~n"),
    Server = spawn(fun() -> server(Main, Verbose) end),

    % Spawn client processes
    io:format("Starting ~p clients to send ~p messages each...~n", [NumClients, NumMessages]),
    [spawn(fun() -> client(Server, NumMessages, Main, Verbose) end) || _ <- lists:seq(1, NumClients)],

    % Wait for all clients and the server to finish
    wait_for_processes(Server, NumClients, Verbose),

    io:format("Terminating.~n"),
    halt(0). % Shutdown the system

% Parses command-line arguments and converts them to integers
parse_args([NumClientsStr, NumMessagesStr | Rest]) ->
    case {string:to_integer(NumClientsStr), string:to_integer(NumMessagesStr)} of
        {{IntClients, _}, {IntMessages, _}} ->
            Verbose = lists:member("verbose", Rest), % Check for verbosity flag
            {ok, IntClients, IntMessages, Verbose};
        _ ->
            {error, "Invalid arguments."}
    end;
parse_args(_) ->
    {error, "Invalid number of arguments."}.

% Client process: Sends messages to the server and waits for responses
client(_, 0, Main, true) ->
    io:format("Client ~p reached max pings, terminating.~n", [self()]),
    Main ! {client, self(), done}; % Notify the main process that client is done
client(_, 0, Main, false) ->
    Main ! {client, self(), done}; % Notify the main process that client is done
client(Server, NumMessages, Main, true) -> % Verbose mode
    % Send a ping to the server
    Server ! {self(), ping},

    % Wait for a pong message
    receive
        pong ->
            io:format("Client ~p received: pong~n", [self()]),
            client(Server, NumMessages - 1, Main, true) % Recursively send more messages
    end;
client(Server, NumMessages, Main, false) -> % Silent mode
    % Send a ping to the server
    Server ! {self(), ping},

    % Wait for a pong message
    receive
        pong ->
            client(Server, NumMessages - 1, Main, false) % Recursively send more messages
    end.

% Server process: Handles ping messages and responds with pong messages
server(Main, true) -> % Verbose mode enabled
    receive
        {From, ping} ->
            io:format("Server ~p received: ping~n", [self()]),
            From ! pong, % Respond with pong
            server(Main, true); % Recursively wait for more messages
        stop ->
            io:format("Server ~p terminating.~n", [self()]),
            Main ! {server, self(), done} % Notify the main process
    end;
server(Main, false) -> % Silent mode
    receive
        {From, ping} ->
            From ! pong, % Respond with pong
            server(Main, false); % Recursively wait for more messages
        stop ->
            Main ! {server, self(), done} % Notify the main process
    end.

% Waiting loop for all clients and the server to signal completion
wait_for_processes(Server, 0, true) -> % Terminate server when clients are done
    io:format("All clients finished, terminating server.~n"),
    Server ! stop,
    receive
        {server, Pid, done} ->
            io:format("Server ~p finished.~n", [Pid])
    end;
wait_for_processes(Server, 0, false) -> % Terminate server when clients are done
    io:format("All clients finished, terminating server.~n"),
    Server ! stop,
    receive
        {server, _Pid, done} -> done
    end;
wait_for_processes(Server, NumClients, true) ->
    receive
        {client, Pid, done} ->
            io:format("Client ~p finished.~n", [Pid]),
            wait_for_processes(Server, NumClients - 1, true) % Decrement client count
    end;
wait_for_processes(Server, NumClients, false) ->
    receive
        {client, _Pid, done} ->
            wait_for_processes(Server, NumClients - 1, false) % Decrement client count
    end.

File: src/native/ping_pong_caf/ping_pong.cpp
/*
 * ping_pong.cpp
 *
 * This source file implements a distributed message-passing system using the
 * Actor model in the CAF framework. It simulates clients sending messages
 * (pings) to the server, which processes messages and
 * responds with pongs. The main process ensures orderly termination.
 *
 * Parameters:
 * - clients: Number of client processes.
 * - messages: Messages each client sends.
 * - verbose: Verbose mode for debugging.
 *
 * The system terminates when all clients and the server finish processing.
 */

#include <caf/actor_system.hpp>
#include <caf/all.hpp>
#include <caf/atom.hpp>
#include <caf/stateful_actor.hpp>
#include <cstddef>
#include <iostream>

using namespace caf;

// Server behavior: Processes "ping" messages from clients and responds with
// "pong". The server notifies the main actor upon termination.
behavior server_fun(event_based_actor *self, const actor &main,
                    const bool verbose) {
  return {[=](const actor &client, atom_value msg) {
            if (msg == atom("ping")) {
              if (verbose)
                aout(self) << "Server <" << self->id()
                           << "> received: " << to_string(msg) << std::endl;
              self->send(client, atom("pong"));
            }
          },
          [=](atom_value msg) {
            if (msg == atom("stop")) {
              if (verbose)
                aout(self) << "Server <" << self->id() << "> terminating."
                           << std::endl;
              self->send(main, atom("server"), self->id(), atom("done"));
              self->quit();
            }
          }};
}

// Client behavior: Sends a predefined number of "ping" messages to the
// server and waits for "pong" responses. It notifies the main actor upon
// completion.
behavior client_fun(event_based_actor *self, const actor &server,
                    const size_t num_messages, const actor &main,
                    const bool verbose) {
  size_t responses_received = 0;

  return {[=](atom_value msg) mutable {
    if (msg == atom("start")) {
      self->send(server, self, atom("ping"));
    }
    if (msg == atom("pong")) {
      ++responses_received;
      if (verbose)
        aout(self) << "Client <" << self->id()
                   << "> received: " << to_string(msg) << std::endl;
      if (responses_received == num_messages) {
        if (verbose)
          aout(self) << "Client <" << self->id()
                     << "> reached max pings, terminating." << std::endl;
        self->send(main, atom("client"), self->id(), atom("done"));
        self->quit();
      } else {
        self->send(server, self, atom("ping"));
      }
    }
  }};
}

// Main actor state to store the server actor and the number of finished clients
struct main_state {
  actor server;
  size_t clients_finished = 0;
};

// Main actor: Waits for all clients and the server to finish before terminating
// the system.
behavior main_fun(stateful_actor<main_state> *self, const size_t num_clients,
                  const bool verbose) {

  return {[=](const actor &server) mutable { self->state.server = server; },

          [=](atom_value type, actor_id id, atom_value status) mutable {
            if (status == atom("done")) {
              if (type == atom("client")) {
                ++self->state.clients_finished;
                if (verbose) {
                  aout(self) << "Client <" << id << "> finished." << std::endl;
                }
                if (self->state.clients_finished == num_clients) {
                  aout(self) << "All clients finished, terminating server."
                             << std::endl;
                  self->send(self->state.server, atom("stop"));
                }
              } else if (type == atom("server")) {
                if (verbose) {
                  aout(self) << "Server <" << id << "> finished." << std::endl;
                }
                aout(self) << "Terminating." << std::endl;
                self->quit();
              }
            }
          }};
}

// Configuration structure for setting up the actor system parameters
struct config : public actor_system_config {
  size_t num_clients = 4;
  size_t num_messages = 10;
  bool verbose = false;

  config() {
    opt_group{custom_options_, "global"}
        .add(num_clients, "clients", "Number of Clients")
        .add(num_messages, "messages", "Number of Messages")
        .add(verbose, "verbose", "Verbose output");
  }
};

// Main function that initializes the actor system and spawns actors
void caf_main(actor_system &sys, const config &cfg) {
  // Spawn main actor to oversee termination
  auto main_actor = sys.spawn(main_fun, cfg.num_clients, cfg.verbose);

  // Spawn server actor
  std::cout << "Starting server..." << std::endl;
  auto server_actor = sys.spawn(server_fun, main_actor, cfg.verbose);
  anon_send(main_actor, server_actor);

  // Spawn client actors
  std::cout << "Starting " << cfg.num_clients << " clients to send "
            << cfg.num_messages << " messages each..." << std::endl;
  for (size_t i = 0; i < cfg.num_clients; ++i) {
    auto client_actor = sys.spawn(client_fun, server_actor, cfg.num_messages,
                                  main_actor, cfg.verbose);
    anon_send(client_actor, atom("start"));
  }
}

// Entry point for the CAF-based actor system
CAF_MAIN()
